lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp

```diff
@@ -758,57 +758,62 @@ bool HandshakeBuilder::visitHandshake(handshake::BranchOp op) {
-                                              /*withElseRegion=*/true);
-
-  // When control signal is true, the first branch is selected
-  rewriter.setInsertionPointToStart(&branchWhenOp.thenRegion().front());
-  rewriter.create<ConnectOp>(insertLoc, result0Valid, argValid);
-  rewriter.create<ConnectOp>(insertLoc, argReady, result0Ready);
+  Value trueResultValid = trueResultSubfields[0];
+  Value trueResultReady = trueResultSubfields[1];
+  Value falseResultValid = falseResultSubfields[0];
+  Value falseResultReady = falseResultSubfields[1];
 
-  if (!op.isControl()) {
-    Value argData = argSubfields[2];
-    Value result0Data = result0Subfields[2];
-    rewriter.create<ConnectOp>(insertLoc, result0Data, argData);
-  }
+  auto conditionArgValid = rewriter.create<AndPrimOp>(
+      insertLoc, conditionValid.getType(), conditionValid, argValid);
+
+  auto conditionNot = rewriter.create<NotPrimOp>(
+      insertLoc, conditionData.getType(), conditionData);
 
-  auto control0ReadyOp = rewriter.create<AndPrimOp>(
-      insertLoc, argValid.getType(), argValid, result0Ready);
-  rewriter.create<ConnectOp>(insertLoc, controlReady, control0ReadyOp);
+  // Connect valid signal of both results.
+  rewriter.create<ConnectOp>(
+      insertLoc, trueResultValid,
+      rewriter.create<AndPrimOp>(insertLoc, conditionData.getType(),
+                                 conditionData, conditionArgValid));
 
-  // When control signal is false, the second branch is selected
-  rewriter.setInsertionPointToStart(&branchWhenOp.elseRegion().front());
-  rewriter.create<ConnectOp>(insertLoc, result1Valid, argValid);
-  rewriter.create<ConnectOp>(insertLoc, argReady, result1Ready);
+  rewriter.create<ConnectOp>(
+      insertLoc, falseResultValid,
+      rewriter.create<AndPrimOp>(insertLoc, conditionNot.getType(),
+                                 conditionNot, conditionArgValid));
 
+  // Connect data signal of both results if applied.
   if (!op.isControl()) {
     Value argData = argSubfields[2];
-    Value result1Data = result1Subfields[2];
-    rewriter.create<ConnectOp>(insertLoc, result1Data, argData);
+    Value trueResultData = trueResultSubfields[2];
+    Value falseResultData = falseResultSubfields[2];
+    rewriter.create<ConnectOp>(insertLoc, trueResultData, argData);
+    rewriter.create<ConnectOp>(insertLoc, falseResultData, argData);
   }
 
-  auto control1ReadyOp = rewriter.create<AndPrimOp>(
-      insertLoc, argValid.getType(), argValid, result1Ready);
-  rewriter.create<ConnectOp>(insertLoc, controlReady, control1ReadyOp);
+  // Connect ready signal of input and condition.
+  auto selectedResultReady = rewriter.create<MuxPrimOp>(
+      insertLoc, trueResultReady.getType(), conditionData, trueResultReady,
+      falseResultReady);
+
+  rewriter.create<ConnectOp>(
+      insertLoc, argReady,
+      rewriter.create<AndPrimOp>(insertLoc, selectedResultReady.getType(),
+                                 selectedResultReady, conditionValid));
```

Comment:
Should the `argReady` signal depend only on `selectedResultReady` and `conditionValid`? What if the condition is valid, but the arg is not valid yet? I think in this case, the arg would be marked ready when it shouldn't, but I may be missing something.

Reply:
Related, I think the logic `selectedResultReady & conditionArgValid` could be used for both `argReady` and `conditionReady`. Does that make sense? I'm looking at Figure 5 [here](http://www.cs.columbia.edu/~sedwards/papers/edwards2019compositional.pdf). I think that would address my correctness concern as well as simplify this be removing one `AndPrimOp` from the current implementation.

Reply:
Make sense. I've updated the implementation.